pylinalg/matrix.py

```diff
@@ -673,6 +673,242 @@ def mat_look_at(eye, target, up_reference, /, *, out=None, dtype=None):
+        n12 * n24 * n41
+        - n14 * n22 * n41
+        + n14 * n21 * n42
+        - n11 * n24 * n42
+        - n12 * n21 * n44
+        + n11 * n22 * n44
+    ) * det_inv
+    oe[14] = (
+        n13 * n22 * n41
+        - n12 * n23 * n41
+        - n13 * n21 * n42
+        + n11 * n23 * n42
+        + n12 * n21 * n43
+        - n11 * n22 * n43
+    ) * det_inv
+
+    oe[3] = t14 * det_inv
+    oe[7] = (
+        n13 * n24 * n31
+        - n14 * n23 * n31
+        + n14 * n21 * n33
+        - n11 * n24 * n33
+        - n13 * n21 * n34
+        + n11 * n23 * n34
+    ) * det_inv
+    oe[11] = (
+        n14 * n22 * n31
+        - n12 * n24 * n31
+        - n14 * n21 * n32
+        + n11 * n24 * n32
+        + n12 * n21 * n34
+        - n11 * n22 * n34
+    ) * det_inv
+    oe[15] = (
+        n12 * n23 * n31
+        - n13 * n22 * n31
+        + n13 * n21 * n32
+        - n11 * n23 * n32
+        - n12 * n21 * n33
+        + n11 * n22 * n33
+    ) * det_inv
+
+    return out
+
+
+if int(np.__version__.split(".")[0]) >= 2:
+    _default_mat_inv_method = "numpy"
+else:
+    _default_mat_inv_method = "manual"
+
+
+def mat_inverse(
+    matrix, /, *, method=_default_mat_inv_method, dtype=None, out=None
+) -> np.ndarray:
+    """
+    Compute the inverse of a matrix.
+
+    Parameters
+    ----------
+    matrix : ndarray, [4, 4]
+        The matrix to invert.
+    method : str, optional
+        The method to use for inversion. The default is "numpy" when
+        numpy version is 2.0.0 or newer, otherwise "manual".
+    dtype : data-type, optional
+        Overrides the data type of the result.
+    out : ndarray, optional
+        A location into which the result is stored. If provided, it must have a
+        shape that the inputs broadcast to. If not provided or None, a
+        freshly-allocated array is returned. A tuple must have length equal to
+        the number of outputs.
+
+    Returns
+    -------
+    out : ndarray, [4, 4]
+        The inverse of the input matrix.
+
+    Notes
+    -----
+    The default method is "numpy" when numpy version >= 2.0.0,
+    which uses the `numpy.linalg.inv` function.
+    The alternative method is "manual", which uses a manual implementation of
+    the inversion algorithm. The manual method is used to avoid a performance
+    issue with `numpy.linalg.inv` on some platforms when numpy version < 2.0.0.
+    See: https://github.com/pygfx/pygfx/issues/763
+    The manual method is slower than the numpy method, but it is guaranteed to work.
+
+    When the matrix is singular, it will return a matrix filled with zeros,
+    This is a common behavior in real-time graphics applications.
+
+    """
+
+    fn = {
+        "numpy": np.linalg.inv,
+        "manual": _mat_inv,
+    }[method]
+
+    matrix = np.asarray(matrix)
+    if out is None:
+        out = np.empty_like(matrix, dtype=dtype)
+    try:
+        out[:] = fn(matrix)
+    except np.linalg.LinAlgError:
+        out[:] = 0
+    return out
```

Comment:
I think with this strategy we can avoid one matrix allocation in the most common use cases:

```suggestion
    matrix = np.asarray(matrix)
    try:
        inverse = fn(matrix)
    except np.linalg.LinAlgError:
        inverse = np.zeros_like(matrix, dtype=dtype)
    if out is None:
        if dtype is not None:
            return inverse.astype(dtype, copy=False)
        return inverse
    else:
        out[:] = inverse
    return out
```